sql/core/src/main/scala/org/apache/spark/sql/execution/HiveResult.scala

```diff
@@ -73,7 +73,13 @@ object HiveResult {
   }
 
   private def zoneId = DateTimeUtils.getZoneId(SQLConf.get.sessionLocalTimeZone)
-  private def dateFormatter = DateFormatter(zoneId)
+  // Date formatting does not depend on time zone ID because it converts local days
+  // since the epoch to local date string. Time zone id does matter only in parsing
+  // when the parser has to handle special values like `now`, `yesterday` and etc.
+  // Here, `dateFormatter` is used only for formatting, so, we can initialize it by
+  // any time zone once, for instance, by the current session time zone. And we can
+  // reuse it even when the session time zone might be changed.
+  private val dateFormatter = DateFormatter(zoneId)
```

Comment:
Just in case, can `hiveResultString()` or `toHiveString()` be called from multiple threads in parallel. If so, we cannot use SimpleDateFormat here because it is not thread-safe. In that case, dateFormatter must use `FastDateFormat` because it is thread-safe.

Reply:
Yes, it can be called concurrently (e.g. by concurrent thriftserver queries).

Reply:
ok. I use `FastDateFormat` for `java.sql.Date` which is thread-safe and DateTimeFormatter for `java.time.LocalDate`. The former one is thread-safe as well. Should be fine.